diff --git coreutils-clean/build-aux/config.sub coreutils-workdir/build-aux/config.sub
index 4aaae46..88f6bbd 100755
--- coreutils-clean/build-aux/config.sub
+++ coreutils-workdir/build-aux/config.sub
@@ -2216,6 +2216,8 @@ case $kernel-$os-$obj in
 		;;
 	managarm-mlibc*- | managarm-kernel*- )
 		;;
+	badgeros-mlibc*- )
+		;;
 	windows*-msvc*-)
 		;;
 	-dietlibc*- | -llvm*- | -mlibc*- | -musl*- | -newlib*- | -relibc*- \
diff --git coreutils-clean/gnulib-tests/getlocalename_l-unsafe.c coreutils-workdir/gnulib-tests/getlocalename_l-unsafe.c
index 6a72c81..82da5d1 100644
--- coreutils-clean/gnulib-tests/getlocalename_l-unsafe.c
+++ coreutils-workdir/gnulib-tests/getlocalename_l-unsafe.c
@@ -650,6 +650,8 @@ getlocalename_l_unsafe (int category, locale_t locale)
       else
         /* It's the "C" or "POSIX" locale.  */
         return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
+#elif defined __mlibc__
+      return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
 #elif defined __ANDROID__
       /* Android API level >= 21 */
       struct __locale_t {
diff --git coreutils-clean/lib/getlocalename_l-unsafe.c coreutils-workdir/lib/getlocalename_l-unsafe.c
index 6a72c81..721a586 100644
--- coreutils-clean/lib/getlocalename_l-unsafe.c
+++ coreutils-workdir/lib/getlocalename_l-unsafe.c
@@ -469,6 +469,8 @@ getlocalename_l_unsafe (int category, locale_t locale)
       struct gl_locale_category_t *plc =
         &locale->category[gl_log2_lcmask_to_index (gl_log2_lc_mask (category))];
       return (struct string_with_storage) { plc->name, STORAGE_OBJECT };
+#elif defined __mlibc__
+      return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
 #elif __GLIBC__ >= 2 && !defined __UCLIBC__
       /* Work around an incorrect definition of the _NL_LOCALE_NAME macro in
          glibc < 2.12.
@@ -650,6 +652,8 @@ getlocalename_l_unsafe (int category, locale_t locale)
       else
         /* It's the "C" or "POSIX" locale.  */
         return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
+#elif defined __mlibc__
+      return (struct string_with_storage) { "C", STORAGE_INDEFINITE };
 #elif defined __ANDROID__
       /* Android API level >= 21 */
       struct __locale_t {
